tensorboard/data/server/types.rs

```diff
@@ -0,0 +1,110 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+//! Core simple types.
+
+use std::borrow::Borrow;
+
+/// A step associated with a record, strictly increasing over time within a record stream.
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Copy, Clone)]
+pub struct Step(pub i64);
+
+/// The wall time of a TensorBoard event.
+///
+/// Wall times represent floating-point seconds since Unix epoch. They must be finite and non-NaN.
+#[derive(Debug, PartialEq, PartialOrd, Copy, Clone)]
+pub struct WallTime(f64);
+
+impl WallTime {
+    /// Parses a wall time from a time stamp representing seconds since Unix epoch.
+    ///
+    /// Returns `None` if the given time is infinite or NaN.
+    pub fn new(time: f64) -> Option<Self> {
+        if time.is_finite() {
+            Some(WallTime(time))
+        } else {
+            None
+        }
+    }
+}
+
+// Wall times are totally ordered and have a total equivalence relation, since we guarantee that
+// they are not NaN.
+#[allow(clippy::derive_ord_xor_partial_ord)] // okay because it agrees with `PartialOrd` impl
```

Comment:
Correct: the `PartialOrd` and `Ord` implementations for a given type
must agree, else many operations are undefined. (Not the scary kind of
“undefined behavior”; it’s just a contract violation on basically every
API that cares about orderings.)

This line disables a Clippy lint, whose docs you’ve linked to. I think
that the purpose is to flag code like this:

```
#[derive(PartialOrd)]
struct BackwardInt(i64);

impl Ord for BackwardInt {
    fn cmp(&self, other: &Self) -> Ordering {
        // Our type is ordered backward!
        self.0.cmp(&other.0).reverse()
    }
}
```

This `Ord` implementation is fine by itself, and you might think that
`#[derive(PartialOrd)]` would implement `PartialOrd` for our type by
delegating to its `Ord`. But in fact it would actually delegate to the
`PartialOrd`s of its fields, which would make the `PartialOrd` and `Ord`
inconsistent in this case.

In our case, the derived `PartialOrd` is consistent with the manual
`Ord` (which you can tell since `cmp` just calls `partial_cmp`), so I
suppressed the lint error. We could write it out—

```rust
impl PartialOrd for WallTime {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}
```

—but I didn’t see much reason to do that.

Reply:
Here’s a topical upstream request to teach the lint to recognize this
safe pattern:
<https://github.com/rust-lang/rust-clippy/issues/6219>